### ISSUE TITLE
Suggestion: Implement optional custom error messages for validation parameter attributes

### ISSUE THREAD

**Issue by mklement0** [PowerShell/PowerShell#3630]
The `Validate*` parameter attributes are a wonderful way of enforcing constraints on arguments (parameter values) that go beyond type checking.

Unfortunately, the generic error messages that result when validation fails are often not helpful enough, and inherently they cannot always be.

Therefore, it would be great if a custom error message could be provided as part of the attribute declaration.

Consider this case:

```powershell
> function foo { param([ValidateScript({ $_ -match '^\p{L}\d+' })] $bar) }; foo -bar baz
foo : Cannot validate argument on parameter 'bar'. The " $_ -match '^\p{L}\d+' " validation script for the argument with value "baz" did not return a result of True.
# ...
```

The error message is noisy and unhelpful, and, in the absence of additional information, _cannot_ be more helpful.

A workaround is to have an explicit `Throw` statement in the script block, but that's clumsy and cumbersome.

It would be nice to do something like this instead:

```powershell
> function foo { param([ValidateScript({ $_ -match '^\p{L}\d+' }, 'You must specify a letter followed by one or more digits.')] $bar) }; foo -bar baz
foo : Cannot validate argument on parameter 'bar'. You must specify a letter followed by one or more digits.
# ...
```

Additional considerations:

* Support placeholders such as `{0}` in the message that can access the argument passed as well as the (other) arguments passed to the validation attribute _constructor_.  
(This is already used internally for the generic error messages.)

* Provide a mechanism to _localize_ the messages.

**Comment by powercode** [comment]
Both ValidateScript and ValidatePattern implements an ErrorMessage property. See  #2728.
```
[ValidateScript("$_ % 2", ErrorMessage = "The item '{0}' did not pass validation of script '{1}'")]
[ValidatePattern("\s+", ErrorMessage="The text '{0}' did not pass validation of regex '{1}'")]
```

**Comment by SteveL-MSFT** [comment]
Looks to me like that PR should address this.

**Comment by mklement0** [comment]
Thanks, guys, I wasn't aware of that.

It's a great step in the right direction, but I think it should be implemented for _all_ validation attributes: see #3745
